src/coreclr/gc/gc.cpp

```diff
@@ -50861,20 +50866,31 @@ CFinalize::RegisterForFinalization (int gen, Object* obj, size_t size)
             return false;
         }
     }
+
+    //          | loop:           /->-->-\   /->->-\           |
+    //          |                /        \ /       \          |
+    // Start:   | ... | dest   |a  other  |b  gen0  |_  free   |  ...
+    //          |              /                               |
+    //     final store: o -->-/                                |
+    //          |                                              |
+    // Result:  | ... | dest    o| other   a| gen0   b| free   |  ...
+
+    assert (dest < FreeListSeg);
     Object*** end_si = &SegQueueLimit (dest);
-    do
+    while (s_i > end_si)
     {
+        assert (s_i > &SegQueue(0));
```

Comment:
will be `>= &SegQueueLimit(0)` once merged with cleanup PR

(The issue is that &SegQueue(0) returns the address of m_Array, which isn't in m_FillPointers and therefore isn't appropriate for pointer comparisons,)

Reply:
(deleted) - It's more complicated than that, and a (separate) rewrite to use SegQueue and SegQueueLimit throughout the loop will be better.